vault/resource_plugin.go

```diff
@@ -0,0 +1,226 @@
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			fieldEnv: {
+				Type:        schema.TypeList,
+				Description: "List of additional environment variables to run the plugin with in KEY=VALUE form.",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Sensitive: true,
+			},
+			fieldOCIImage: {
+				Type:        schema.TypeString,
+				Description: "OCI image to run. If specified, setting command, args, and env will update the container's entrypoint, args, and environment variables (append-only) respectively.",
+				Optional:    true,
+			},
+			fieldRuntime: {
+				Type:        schema.TypeString,
+				Description: "Vault plugin runtime to use if oci_image is specified.",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func pluginWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	pluginType, err := api.ParsePluginType(d.Get(consts.FieldType).(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	name := d.Get(consts.FieldName).(string)
+	version := d.Get(consts.FieldVersion).(string)
+	id := fmt.Sprintf("%s/%s", pluginType, name)
+	if version != "" {
+		id = fmt.Sprintf("%s/%s", id, version)
+	}
+
+	log.Printf("[DEBUG] Writing plugin %q", id)
+	err = client.Sys().RegisterPluginWithContext(ctx, &api.RegisterPluginInput{
+		Type:     pluginType,
+		Name:     name,
+		Version:  version,
+		SHA256:   d.Get(fieldSHA256).(string),
+		Command:  d.Get(fieldCommand).(string),
+		Args:     util.ToStringArray(d.Get(fieldArgs).([]interface{})),
+		Env:      util.ToStringArray(d.Get(fieldEnv).([]interface{})),
+		OCIImage: d.Get(fieldOCIImage).(string),
+		Runtime:  d.Get(fieldRuntime).(string),
+	})
+	if err != nil {
+		return diag.Errorf("error updating plugin %q: %s", id, err)
+	}
+	log.Printf("[DEBUG] Wrote plugin %q", id)
+
+	d.SetId(id)
+
+	return nil
+}
+
+func pluginRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	var typ, name, version string
+	parts := strings.Split(d.Id(), "/")
+	switch len(parts) {
+	case 2:
+		typ, name = parts[0], parts[1]
+	case 3:
+		typ, name, version = parts[0], parts[1], parts[2]
+	default:
+		return diag.Errorf("invalid ID %q, must be of form <type>/<name> or <type>/<name>/<semantic-version>", d.Id())
+	}
+
+	pluginType, err := api.ParsePluginType(typ)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp, err := client.Sys().GetPluginWithContext(ctx, &api.GetPluginInput{
+		Type:    pluginType,
+		Name:    name,
+		Version: version,
+	})
+
+	if err != nil {
+		return diag.Errorf("error reading plugin %q %q %q: %s", pluginType, name, version, err)
+	}
+
+	if err := d.Set(consts.FieldType, typ); err != nil {
```

Comment:
nit: consolidating repeated code into a single loop. 
```suggestion
        for _, k := range pluginFields {
		if v, ok := resp.Data[k]; ok {
			if err := d.Set(k, v); err != nil {
				return diag.Errorf("error setting %s for Registered Plugin %q: %q", k, path, err)
			}
		}
	}
```